crates/flux-fhir-analysis/src/lib.rs

```diff
@@ -137,6 +138,24 @@ fn adt_def(genv: GlobalEnv, def_id: LocalDefId) -> QueryResult<rty::AdtDef> {
     Ok(rty::AdtDef::new(adt_def, genv.adt_sort_def_of(def_id)?, invariants, is_opaque))
 }
 
+fn constant_info(genv: GlobalEnv, local_def_id: LocalDefId) -> QueryResult<rty::ConstantInfo> {
+    let def_id = genv.maybe_extern_id(local_def_id);
+    let node = genv.map().node(def_id.local_id())?;
+    let owner = rustc_hir::OwnerId { def_id: local_def_id };
+    let constant = if let fhir::Node::Item(item) = node
+        && let ItemKind::Const(constant) = &item.kind
+    {
+        constant
+    } else {
+        &fhir::ConstantInfo { owner, expr: None }
+    };
+    let Some(sort) = genv.sort_of_def_id(owner.def_id.to_def_id()).emit(&genv)? else {
+        return Ok(rty::ConstantInfo::Uninterpreted);
+    };
+    let wfckresults = wf::check_constant(genv, owner, constant, &sort)?;
+    conv::conv_constant(genv, local_def_id.to_def_id(), constant, &wfckresults)
```

Comment:
I'd make this more explicit about what stuff we support (I'm assuming the expr is directly stored in `ItemKind::Const`)

```suggestion
    let def_id = genv.maybe_extern_id(local_def_id);
    let node = genv.map().node(def_id.local_id())?;
    match node {
        fhir::Node::Item { kind: ItemKind::Const(Some(expr)) } => {
            let Some(sort) = genv.sort_of_def_id(owner.def_id.to_def_id()).emit(&genv)? else {
                return Ok(rty::ConstantInfo::Uninterpreted);
            };
            let wfckresults = wf::check_constant(genv, owner, expr, &sort)?;
            conv::conv_constant(genv, local_def_id.to_def_id(), expr, &wfckresults)
        }
        fhir::Node::TraitItem { kind: TraitItemKind::Const, .. } |
        fhir::Node::ImplItem { kind: ImplItemKind::Const, .. } |
        fhir::Node::Item { kind: ItemKind::Const(None) } |
        fhir::Node::AnonConst => {
            rty::ConstantInfo::Uninterpreted
        }
```

Note that if we have the expr directly in `ItemKind::Const`, `wf::check_constant` and `conv::conv_constant` don't make much sense and we can do everything directly on the expr.

Reply:
Hmm, the above doesn't seem to be valid rust syntax -- the hassle is that `Item(ItemKind)` and `TraitItem(TraitItemKind)` and `ImplItem(ImplItemKind)` and I can't figure out the syntax for matching on the inner (X-Kind)'s `kind` field?

Reply:
odd I had tried the thing you suggested ... maybe rust-analyzer had gotten into one of its funky states...